integration-tests/tests/client/process_blocks.rs

```diff
@@ -3891,4 +3891,111 @@ mod contract_precompilation_tests {
+
+        let network_adapter = Arc::new(MockNetworkAdapter::default());
+        let networks = vec![network_adapter.clone(); num_clients];
+        let mut env = TestEnv::new_with_runtime_and_network_adapter(
+            ChainGenesis::test(),
+            num_clients,
+            genesis.config.num_block_producer_seats,
+            runtime_adapters,
+            networks,
+        );
+
+        let genesis_block = env.clients[0].chain.get_block_by_height(0).unwrap().clone();
+        assert_eq!(genesis_block.header().height(), 0);
+
+        // singer to send money from test0 to test1
+        let signer = InMemorySigner::from_seed("test0".parse().unwrap(), KeyType::ED25519, "test0");
+        let mut prev_block_hash = genesis_block.hash().clone();
+        let epoch_id = genesis_block.header().epoch_id().clone();
+        for next_block_height in 1..=10 {
+            let tx = SignedTransaction::send_money(
+                1,
+                "test0".parse().unwrap(),
+                "test1".parse().unwrap(),
+                &signer,
+                10000000,
+                prev_block_hash,
+            );
+
+            let block_producer_account = env.clients[0]
+                .runtime_adapter
+                .get_block_producer(&epoch_id, next_block_height as u64)
+                .unwrap();
+            let producer_id = accounts.iter().position(|r| r == &block_producer_account).unwrap();
+
+            // every client should know tha mapping between account id and shard id
+            let shard_id = env.clients[0]
+                .runtime_adapter
+                .account_id_to_shard_id(&signer.account_id, &epoch_id)
+                .unwrap();
+            // every client should know who is next chunk producer
+            let chunk_producer_account = env.clients[0]
+                .runtime_adapter
+                .get_chunk_producer(&epoch_id, next_block_height as u64, shard_id)
+                .unwrap();
+            let chunk_id = accounts.iter().position(|r| r == &chunk_producer_account).unwrap();
+            {
+                let mut client = &mut env.clients[chunk_id];
+                let _r = client.process_tx(tx, false, false);
+                let (encoded_chunk, merkle_paths, receipts) =
+                    create_chunk_on_height(&mut client, next_block_height as u64);
+                assert_eq!(receipts.len(), 0);
+                let mut chain_store = ChainStore::new(
+                    client.chain.store().owned_store(),
+                    genesis_block.header().height(),
+                );
+                //distribute chunks
+                client
+                    .shards_mgr
+                    .distribute_encoded_chunk(
+                        encoded_chunk,
+                        merkle_paths,
+                        receipts,
+                        &mut chain_store,
+                    )
+                    .unwrap();
```

Comment:
This is certainly moving in the right direction. However, I am not sure that it is correct. Here we distribute the chunk parts but I don't see any code that enables other block producers to receive them. You might have to change network adapter to make sure that chunk parts are delivered.